### PR TITLE
ThreadStats - Send correct metric type instead of always sending Gauge

### PR DESCRIPTION
**Pull request by nilabhsagar**
# ThreadStats - Send correct Metric Type instead of always sending Gauge

**Problem**
At present in ThreadStats various types of metric gets created like gauge, count, histogram, etc. but at the time of flushing for each metric type is set to "gauge". This is a problem as sending count from server becomes gauge in the web. Changing the type in web will not have any effect as from the server it is always coming as gauge.

**Solution**
Assign the right metric type based on the metric functionality.

### CODE REVIEW

**Comment by nilabhsagar** [comment]
The above pull request will fix the issue #245

**Comment by jayshao** [comment]
@nilabhsagar looks like this failed the Travis build for Python 2.5, but only for line-length. Are you still looking at this?

**Comment by nilabhsagar** [comment]
@jayshao Yes, I can fix it. Could you help me with the exact line number to fix.

**Comment by nilabhsagar** [comment]
@jayshao Its fixed now, All Travis build passed.

**Comment by nilabhsagar** [comment]
Hi, Any plan to review and merge with the master?

**Comment by nilabhsagar** [comment]
@yannmh Request you to review.

**Comment by nicsnoek** [comment]
Please merge ASAP. Thanks @nilabhsagar for fixing this enormity ;-)

**Comment by irabinovitch** [comment]
@niamster Thanks for reaching out.  We have been looking into your PR, and unfortunately it is not mergeable as is.  Currently the server side API would also require changes to support some of these metric types.  I expect we'll be able to provide more insight into viability and/or timeline next week.

**Comment by nilabhsagar** [comment]
@irabinovitch Sure, Let me know when some steps are required from my side.

**Comment by yannmh** [comment]
Thanks for working on a fix @nilabhsagar 🙇  I apologize for the late reply.

Your changes look good, and so do the tests. 
In order to merge the fix, we'd need to make two additional changes:

1. Metrics type `count` (or `rate`) require an additional `interval` parameter. It instruments the backend to update the metric metadata accordingly and makes it able to perform time rollup operations (such as `as_rate()`, `as_count`...). 
  The `flush` method can be updated to account for this extra parameter, similar to https://github.com/DataDog/dd-agent/blob/master/aggregator.py#L208-L222.
2. We encourage to use the metric type `rate` instead of `count`. Both types are very similar (two representations of the same information), but the `rate` type is generally more used at Datadog, hence more tested. This can be performed simply by dividing the value by the interval as in https://github.com/DataDog/dd-agent/blob/master/aggregator.py#L210.

Please let me know if you'd like make these changes, or have us add them to your existing PR.

**Comment by nilabhsagar** [comment]
@yannmh I have done the changes based on your suggestion. I hope it is aligned with your suggestion. Do let me know if I am missing something. 

Please verify the commit "7cb37c9" as others are just line wrapping issue fix.

File: datadog/threadstats/metrics.py [hunk 1]
```diff
@@ -55,9 +57,9 @@ def __init__(self, name, tags, host):
     def add_point(self, value):
         self.count.append(value)
 
-    def flush(self, timestamp):
+    def flush(self, timestamp, interval):
         count = sum(self.count, 0)
-        return [(timestamp, count, self.name, self.tags, self.host)]
+        return [(timestamp, count, self.name, self.tags, self.host, MetricType.Rate, interval)]
```
Comment (yannmh) [inline datadog/threadstats/metrics.py]: We want to replace `count` by `count/float(interval)` to report a proper rate.

File: datadog/threadstats/metrics.py [hunk 2]
```diff
@@ -88,21 +90,26 @@ def add_point(self, value):
...
         metrics = [
-            (timestamp, self.min, '%s.min' % self.name, self.tags, self.host),
-            (timestamp, self.max, '%s.max' % self.name, self.tags, self.host),
-            (timestamp, self.count, '%s.count' % self.name, self.tags, self.host),
-            (timestamp, self.average(), '%s.avg' % self.name, self.tags, self.host)
+            (timestamp, self.min, '%s.min' % self.name,
+             self.tags, self.host, MetricType.Gauge, interval),
+            (timestamp, self.max, '%s.max' % self.name,
+             self.tags, self.host, MetricType.Gauge, interval),
+            (timestamp, self.count, '%s.count' % self.name,
```
Comment (yannmh) [inline datadog/threadstats/metrics.py]: Same as above. We need to transform the count to a proper rate.

**Comment by yannmh** [comment]
@nilabhsagar added a small comment.

Other than that, it looks good!

**Comment by yannmh** [comment]
Thank you @nilabhsagar. This will be part of our next release, which will go out this week.

**Pull request merged**